### PR TITLE
BackgroundSweeperImpl shutdown and fatal metrics

### PR DESCRIPTION
**Pull request by tboam**
Added 2 new sweep outcomes to the BackgroundSweeperImpl for permanent end states.

**Goals (and why)**: We want to be able to detect and immediately alarm if the Background Sweeper fails.

**Implementation Description (bullets)**:  Add 2 new SweepOutcome values to the enum.  These differ to previous values in that they are permanent end states so we record and publish the metric differently: rather than use a reservoir which evicts values we permanently switch modes for these metric types.

This is done in a bit of a brute force manner for now but there are only 2 of these permanent outcomes so don't think it warrants special handling (if we add more then I'd change the enum value to know if it is permanent or not and store non-permanent values in the reservoir and permanent values in a separate Map).

Set the shutdown flag either if the shutdown method is called on the sweeper or the thread is interrupted directly as either could technically happen.

**Concerns (what feedback would you like?)**: Have I introduced a bug?  There's no testing at the top level of the sweeper.

**Where should we start reviewing?**: Single class.

**Priority (whenever / two weeks / yesterday)**: 1 week.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2884)

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java [hunk 1]
```diff
@@ -271,6 +277,7 @@ SweepLocks createSweepLocks() {
 
     @Override
     public synchronized void shutdown() {
+        sweepOutcomeMetrics.shutdown();
```
Comment (hsaraogi) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: We should call `sweepOutcomeMetrics .registerOccurrenceOf(SHUTDOWN);`

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java [hunk 2]
```diff
@@ -131,6 +132,11 @@ public void run() {
             }
         } catch (InterruptedException e) {
             log.warn("Shutting down background sweeper. Please restart the service to rerun background sweep.");
+            sweepOutcomeMetrics.shutdown();
```
Comment (hsaraogi) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: We should call `sweepOutcomeMetrics .registerOccurrenceOf(SHUTDOWN);`

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java [hunk 3]
```diff
@@ -131,6 +132,11 @@ public void run() {
...
+        } catch (Throwable t) {
+            log.error("BackgroundSweeper failed fatally and will not rerun until restarted: {}",
+                    UnsafeArg.of("message", t.getMessage()), t);
+            sweepOutcomeMetrics.fatal();
```
Comment (hsaraogi) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: We should call `sweepOutcomeMetrics .registerOccurrenceOf(FATAL);`

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java [hunk 4]
```diff
@@ -286,29 +293,59 @@ public synchronized void shutdown() {
...
+            }
+            if (outcome == SweepOutcome.FATAL) {
+                fatal();
+                return;
+            }
+
             reservoir.update(outcome.ordinal());
         }
+
+        public void shutdown() {
```
Comment (hsaraogi) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: We should inline `shutdown()` and `fatal()` as these new APIs to SweepOutcomeMetrics could be confusing. I see this as something I register my outcome with and then it reports the last outcome when I call `getOutcome`.

Comment (tboam) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: Yeah, I wrote this the other way round - `shutdown()` and `fatal()` were separate things so it made sense, but then I reused the SweepOutcome to make the metric creation easy and so had to guard against someone passing them to `registerOccurenceOf()`.

I'll inline everything as it should reduce API size.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java [hunk 5]
```diff
@@ -286,28 +293,50 @@ public synchronized void shutdown() {
...
+        NOT_ENOUGH_DB_NODES_ONLINE, TABLE_DROPPED_WHILE_SWEEPING, ERROR,
+        SHUTDOWN, FATAL
     }
 
     private class SweepOutcomeMetrics {
         private final MetricsManager metricsManager = new MetricsManager();
         private final SlidingTimeWindowReservoir reservoir;
 
+        private boolean shutdown;
+        private boolean fatal;
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: Should this be volatile?

Comment (tboam) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: As discussed I think we're ok with an update taking some amount of time before it is read by the metrics thread.  We only write in one driection and it is only a boolean so we won't get weird race conditions

**Pull request merged**